### PR TITLE
Mcp server pagination

### LINKED ISSUE #33711: MCP: Add pagination for `tools/list`
When listing tools, the MCP spec allows for [pagination](https://modelcontextprotocol.io/specification/2025-11-25/server/utilities/pagination) so that if there are a large number of tools, all the information can be split across several requests rather than returned all in one request.

Pagination is controlled by the server, so we decide how many results to return, and what the cursor is.

Probably the easiest thing to do would be to make the cursor the name of the last tool returned. As long as the list of tools remains the same, if the user provides a cursor, we go down the list until we find the corresponding tool and then start our result from the next one.

We will need to pick a page size. 20 seems reasonable as a starting point. In the future we might make it configurable, but I don't think we need to do that now.

We will need to update existing tests which call `tools/list` which receive a result larger than our page size. We can avoid changing the existing tests too much by making multiple requests and then sticking the results together. I think it would make sense to add a method to `McpClient` to call `tools/list` as many times as necessary and collect all the results.

### PR DESCRIPTION
**Pull request by ttmunyunguma**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".
Resolves #33711

### CODE REVIEW

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java [hunk 1]
```diff
@@ -278,17 +281,50 @@ private void cleanup(ExecutionRequestId requestId) {
...
+                if (count >= PAGE_SIZE) {
+                    if (i + 1 < allTools.size()) {
+                        nextCursor = tmd.name();
+                    }
+                    break;
                 }
             }
-            ToolResult toolResult = new ToolResult(response);
-            transport.sendResponse(toolResult);
         }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: A slight weirdness here: if there are exactly `PAGE_SIZE` tools to return but there are more tools later in the list that the user is not authorized for, we will return a cursor, but calling with that cursor will return no results.

Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: This whole bit feels like it should be a much simpler stream operation, but it's awkward to discover whether there are more results to decide whether we need a cursor or not so I'm not sure whether that's possible.

I do wonder whether it's worth abstracting out this pagination logic at this point so that it can be tested easily and re-used elsewhere.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolRegistry.java [hunk 2]
```diff
@@ -51,8 +52,14 @@ public boolean hasTools() {
...
-        return new ArrayList<>(tools.values());
+    /**
+     * List of tools sorted by tool name in alphabetical order. This is useful for
+     * pagination
+     */
+    public List<ToolMetadata> getAllTools() {
+        List<ToolMetadata> result = new ArrayList<>(tools.values());
+        result.sort(Comparator.comparing(ToolMetadata::name));
+        return result;
     }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolRegistry.java:63]: In the long term, I'm unsure whether we want the tools sorted by name (one issue is that if tools are added while the server is running, the user may miss them when retrieving multiple pages of tools, quarkus solves this by ensuring that new tools are added at the end).

I also don't like sorting the list every time we retrieve it, but that's only really going to be an issue if there are a large number of tools. We can leave this as is for now since I need to work on the ToolRegistry for #33351 anyway.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/utils/McpClient.java [hunk 3]
```diff
@@ -244,4 +246,67 @@ public void callMCPNotification(LibertyServer server,
...
+            JSONArray tools = result.getJSONArray("tools");
+            for (int i = 0; i < tools.length(); i++) {
+                allTools.put(tools.get(i));
+            }
+
+            cursor = result.optString("nextCursor", null);
+            //guard against edge cases where JSON returns cursor="" instead of omitting the cursor
+            if (cursor != null && cursor.isEmpty()) {
+                cursor = null;
+            }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/utils/McpClient.java]: Remove this check, or make it an assertion. We want to guard against such edge cases occurring, not ignore them in our tests.

File: dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops [hunk 4]
```diff
@@ -185,12 +185,19 @@ CWMCM0019E.error.encoding.element.useraction=Review the server message.log and F
...
 CWMCM0021E.defaultvalue.character.invalid.length.explanation=A default value for a Character type argument was provided, but it does not contain exactly one character. Character arguments must have a single character string.
-CWMCM0021E.defaultvalue.character.invalid.length.useraction=Ensure that the default value for an argument with a Character type is a string with exactly one character.
+CWMCM0021E.defaultvalue.character.invalid.length.useraction=Ensure that the default value for an argument with a Character type is string with exactly one character.
+
+# Used for server logging
+# do not translate MCP
+# {0} = The cursor value string
+CWMCM0022E.invalid.cursor.value=CWMCM0022E: The cursor value provided: {0} is invalid
+CWMCM0022E.invalid.cursor.value.explanation=A cursor value was provided as a param to the tool/list command, but the value is invalid or no tool with that name was found
+CWMCM0022E.invalid.cursor.value.useraction=Check that the cursor string value is correct and corresponds to a tool in the tool list
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops]:
```suggestion
# Used as part of a response to a client
# {0} = The cursor value string
CWMCM0022E.invalid.cursor.value=CWMCM0022E: The "{0}" cursor value is invalid.
CWMCM0022E.invalid.cursor.value.explanation=A cursor value was provided as a parameter to the tool/list command, but the value is invalid.
CWMCM0022E.invalid.cursor.value.useraction=Check that the cursor value is provided exactly as it was returned from a previous tool/list call.
```

- The message is returned to the client, not logged in the server log
- We can use the "The _name_ noun" format here. Usually we wouldn't use quotes, but depending on the cursor value it might get confusing so I've used them here.
- Use full sentences
- Cursors are meant to [be opaque tokens to the client](https://modelcontextprotocol.io/specification/2025-11-25/server/utilities/pagination#implementation-guidelines). Although at the moment it is a tool name, there's no requirement that it will always be a tool name so we shouldn't describe it as such in the message.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java [hunk 5]
```diff
@@ -278,17 +281,50 @@ private void cleanup(ExecutionRequestId requestId) {
     private void listTools(McpTransport transport) throws IOException {
         ToolRegistry toolRegistry = ToolRegistry.get();
 
+        if (!toolRegistry.hasTools()) {
+            transport.sendResponse(new ToolResult(List.of()));
+            return;
+        }
+
+        McpToolListParams params = transport.getParams(McpToolListParams.class);
+        String cursor = params != null ? params.getCursor() : null;
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java:288]: This is a bit odd. `params` may be omitted from a request, so having it come back `null` would seem possible, but `transport.getParams` eventually calls `jsonb.toJson(params)` which is [documented to throw a `NullPointerException` if `params` were `null`](https://jakarta.ee/specifications/platform/11/apidocs/jakarta/json/bind/jsonb#toJson(java.lang.Object)).

So either we don't need to check for `null` here because `params` is never `null` (though I'm not sure how that would occur), or `jsonb.toJson` is not throwing an exception when it should (which seems unlikely).

Could you investigate please?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java [hunk 6]
```diff
@@ -278,17 +281,50 @@ private void cleanup(ExecutionRequestId requestId) {
...
+
+        List<ToolMetadata> allTools = new ArrayList<>(toolRegistry.getAllTools());
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: `getAllTools()` already does a copy, we don't need to make another one.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java [hunk 7]
```diff
@@ -278,17 +281,50 @@ private void cleanup(ExecutionRequestId requestId) {
...
         List<ToolDescription> response = new LinkedList<>();
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: `ArrayList` is usually better, particularly when we're not doing lots of insertions at the start or middle of the list.

I would stick to `ArrayList` unless you know you're doing something that `ArrayList` is bad at, or you've profiled it.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/utils/McpClient.java [hunk 8]
```diff
@@ -244,4 +246,67 @@ public void callMCPNotification(LibertyServer server,
...
+
+        JSONObject combinedResult = new JSONObject();
+        combinedResult.put("tools", allTools);
+
+        JSONObject combinedResponse = new JSONObject();
+        combinedResponse.put("jsonrpc", "2.0");
+        combinedResponse.put("id", 1);
+        combinedResponse.put("result", combinedResult);
+
+        return combinedResponse.toString();
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/utils/McpClient.java:309]: It seems a bit unnecessary to recombine all the results into one JSON object, but it does maintain the existing tests, so let's leave it as-is.

**Review by Azquelt — CHANGES_REQUESTED** [review]
In addition to the inline comments, files that have been modified need the copyright date updated.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java (hunk 1 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: I have fixed this bug by first checking for `PAGE_SIZE + 1` authorised tools. This will tell me that there's at least 1 tool to come back for in the next page

Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: I have refactored to used streams, you're right it is much cleaner now. Although I can't figure out reusability yet, still investigating this

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java (hunk 6 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: Fixed, thanks

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java (hunk 7 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: Removed

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java (hunk 5 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java:288]: When `params` is omitted, `this.params` in `McpRequest.getParams`  is null. Following through the process, I saw that `jsonb.toJson(this.params)` when `this.params == null` returns `null` instead of throwing an exception. It could possibly be a certain non-strict json serialization happening here, but I could potentially add a guard here to make sure that if `params == null` we always return a `null`, I do need to keep the null check in `McpServlet` though because when the params is omitted from the request, that value is definitely coming back as null

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/utils/McpClient.java (hunk 3 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/utils/McpClient.java]: I've added it as an assertion here

File: dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops [hunk 9]
```diff
@@ -194,3 +194,10 @@ CWMCM0020E.defaultvalue.conversion.error.useraction=Ensure that the default valu
 CWMCM0021E.defaultvalue.character.invalid.length=CWMCM0021E: A character default value must be exactly one character, but was {0}
 CWMCM0021E.defaultvalue.character.invalid.length.explanation=A default value for a Character type argument was provided, but it does not contain exactly one character. Character arguments must have a single character string.
 CWMCM0021E.defaultvalue.character.invalid.length.useraction=Ensure that the default value for an argument with a Character type is a string with exactly one character.
+
+# Used for server logging
+# do not translate MCP
+# {0} = The cursor value string
+CWMCM0022E.invalid.cursor.value=CWMCM0022E: The "{0}" cursor value is invalid.
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops:201]: Note to ID: usually we wouldn't use quotes around a value, but I think here it might be unclear without the quotes.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/requests/McpRequest.java [hunk 10]
```diff
@@ -66,6 +66,9 @@ public RequestMethod getRequestMethod() {
      * @return the method parameters converted to {@code type}
      */
     public <T> T getParams(Class<T> type, Jsonb jsonb) {
+        if (this.params == null) {
+            return null;
+        }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/requests/McpRequest.java]:
```suggestion
     * @return the method parameters converted to {@code type}, or {@code null} if the request did not include parameters
     */
    public <T> T getParams(Class<T> type, Jsonb jsonb) {
        if (this.params == null) {
            return null;
        }
```

This may have been true before, but let's update the Javadoc to make it explicit that this is what we expect.

**Review by Azquelt — COMMENTED** [review]
There's one bit of Javadoc that should be updated, but everything else looks good.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java (hunk 1 above)
Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: I think the way you've rewritten it is shorter and more concise and we don't need to abstract it out, particularly as we don't have a place we want to re-use it yet.

If we did need to abstract it out, I'd consider an interface like this:

```java
interface Paginator<T> {
    Paginator<T> create(int pageSize, Function<T, String> getCursor);
    
    Page<T> paginate(Collection<T> allItems, String cursor);
    
    record Page<T>(Collection<T>, String cursor) {};
}
```

Depending on what exactly we need to do, we might want to use streams rather than collections in some places or add an additional method to do post-processing of the items, e.g.:

```java
<U> Page<U> paginateWithTransform(Collection<T> allItems, String cursor, Function<T, U> transform);
```

**Comment by Azquelt** [comment]
Also, please do the checks in the PR description and tick the boxes.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java (hunk 1 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: Thanks. If it's all the same with you I'd just stick with the current implementation and let us refactor to the abstraction when we need it

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/requests/McpRequest.java (hunk 10 above)
Comment (ttmunyunguma) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/requests/McpRequest.java]: Done, thanks

**Comment by ttmunyunguma** [comment]
@OpenLiberty/message-reviewer Please review.

- dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops

**Review by Azquelt — DISMISSED** [review]

File: dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops [hunk 11]
```diff
@@ -194,3 +194,10 @@ CWMCM0020E.defaultvalue.conversion.error.useraction=Ensure that the default valu
...
+CWMCM0022E.invalid.cursor.value.explanation=A cursor value was provided as a parameter to the tool/list command, but the value is invalid.
+CWMCM0022E.invalid.cursor.value.useraction=Check that the cursor value is provided exactly as it was returned from a previous tool/list call.
```
Comment (ramkumar-k-9286) [inline dev/io.openliberty.mcp.internal/resources/io/openliberty/mcp/internal/resources/CWMCM.nlsprops]:
```suggestion
CWMCM0022E.invalid.cursor.value.useraction=Ensure that the cursor value is provided exactly as it was returned from the previous tool/list call.
```

**Comment by ttmunyunguma** [comment]
@ramkumar-k-9286 I have committed your suggestion, please re-review

**Comment by Azquelt** [comment]
@OpenLiberty/message-reviewer Could you approve the messages with the applied changes please?

**Review by Azquelt — APPROVED** [review]
